src/db.rs

```diff
@@ -266,6 +267,51 @@ impl DbInner {
         }
     }
 
+    #[allow(clippy::panic)]
+    pub async fn write_batch_with_options(&self, batch: &WriteBatch, options: &WriteOptions) {
+        self.maybe_apply_backpressure().await;
+
+        let current_table = if self.wal_enabled() {
+            let mut guard = self.state.write();
+            let current_wal = guard.wal();
+            for op in &batch.ops {
+                match op {
+                    WriteOp::Put(key, value) => {
+                        current_wal.put(key, value);
+                    }
```

Comment:
if the process crashes between two operations, will atomicity still be guaranteed?

Reply:
> if the process crashes between two operations, will atomicity still be guaranteed?

I believe so. Since we hold the lock for the WAL/memtable, no data can be flushed during the `put` loop. Once the loop is done, the lock will be released. Only then may a flush occur. If the process crashes, the in-memory mutations are lost and no partial batch is written.

I believe the same holds true for situations where we're in a degraded state, as well. Until https://github.com/slatedb/slatedb/issues/111 is done, we don't forward all failures to the client, so we might end up with dead threads. Even in such situations, the active WAL/memtable in-memory structure will be atomically modified.

Reply:
Ok, I just pushed 51071851c0fef34cf1391b061238b5806e521ad3. This code change adds a `spawn_blocking` task in `write_batch_with_options` so that the async runtime threads aren't blocked while we're iterating on large batches. The change also reverts `WriteBatch` back to an owned objet that contains `Vec<[u8]>` so that we can safely move the batch into the `spawn_blocking` closure.

I'm not totally sure about the serializability of multiple atomic writes using this approach. If do three consecutive batch writes, will the async runtime scheduler always guarantee that batch1 will go in before batch2 and batch2 before batch3?

Reply:
> I'm not totally sure about the serializability of multiple atomic writes using this approach. If do three consecutive batch writes, will the async runtime scheduler always guarantee that batch1 will go in before batch2 and batch2 before batch3?

I'm pretty sure there are no such guarantees about task execution order. Why do we care though? At this point the writes are all technically concurrent.

Reply:
I was thinking of the case where a single writer on a single thread does:

```
write_batch(b1);
write_batch(b2);
write_batch(b3);
```

In this case, the user would expect the batches to be written in that order. But if we were spawn blocking, b2 might go before b1 since they're separate threads.

This isn't a problem with the new approach that uses a commit pipeline (see my PR).